Add confidence intervals to the cumulative incidence estimator for competing risks.

mvlvrd opened the pull request:
There are some minimal differences in the results of the confidence intervals with respect to the implementation in
the R package cmprsk, so we add some tolerance to the assertions. Apart from the Aalen estimator used in cmprsk we include two other estimators, but these are not tested so thoroughly, because we couldn't find implementations.

There is a new data set with three different terminal states.



<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://scikit-survival.readthedocs.io/en/latest/contributing.html#making-changes-to-the-code
-->

**Checklist**
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] closes #xxxx
- [x] pytest passes
- [x] tests are included
- [x] code is well formatted
- [x] documentation renders correctly

**What does this implement/fix? Explain your changes**

This adds the missing Confidence Intervals for the cumulative incidence in competing risks.

mvlvrd commented:
Thanks for the reply!
- Aalen estimator is based on counting process theory and it is tested against the `cmprsk` R package, implemented by Gray. The actual formula used is Eq. 4.5 from Pintilie's book.
- Dinse comes from the Delta method applied to the cumulative incidence estimator, assumed it follows a multinomial distribution. I couldn't find any external implementation, so the test data is the data generated by the current code. The formula used is derived from the original Dinse-Larson paper.
- Dinse_approx is a large number of events approximation of the Dinse estimator. It seems that the approximation was popular in the  00's but with current computing power I guess it is not relevant anymore. Nevertheless, I have included it because I couldn't find real data to test the "full" Dinse estimator. The formulas used are derived from Dinse and Larson original paper and are equivalent to Eq. 4.4 in Pintilie's book.
 
There are multiple versions of both kind of estimators. According to Braun and Yuan (Statist. Med. 2007; 26:1170–1180), the Dinse-based estimators seems to work better than the Gray version of the Aalen-based estimator and "are fairly accurate even in small samples". I have decided to include both kind of estimators for completeness and because the Aalen estimators seem to be the standard due to the R implementation.

Apart from this, I have the personal impression that the Aalen-based estimators would not work properly in the case of discrete time because of the possibility of multiple simultaneous events. But I don't understand the theory well enough to justify it.

So in practice, I think Dinse is the best practical solution, while Aalen would be useful when comparing with the `cmprsk`  implementation. Dinse_approx could be deprecated, but may be useful for historical reasons.

I am sorry if this is confusing, I am not an expert on these issues and it took me a while to go through the literature and such.

BR,

mvlvrd commented:
I tried to back-engineer the `cmprsk` code and I think it may be not the same formula. I plotted all three components for both the Fortran code (`va`, `vb` and `vc`) and our implementation (`var_a`, `var_b` and `var_c`) and they are quite different from each other, but somehow the differences cancel out.


Regarding the Dinse formula, I think (B1) in Choudhury (2002) is wrong: 
                   
$$ \prod_{l=1}^{r-1} \frac{c_{(l)}}{N_l(N_l - c_{(l)})} $$

should be
                          
$$ \prod_{l=1}^{r-1} \left( 1 + \frac{c_{(l)}}{N_l(N_l - c_{(l)})} \right) $$

The right formula can be deduced from [Dinse et al. (1986)](https://www.researchgate.net/profile/Gregg-Dinse/publication/243082882_A_Note_on_Semi-Markov_Models_for_Partially_Censored_Data/links/57e0887908ae3f2d793eb293/A-Note-on-Semi-Markov-Models-for-Partially-Censored-Data.pdf?_tp=eyJjb250ZXh0Ijp7ImZpcnN0UGFnZSI6InB1YmxpY2F0aW9uIiwicGFnZSI6InB1YmxpY2F0aW9uIn19) Eq. 5  
I actually first tried to implement it using the formula in Choudhury's paper, but got very bad results. Then I went to the original Dinse's paper and noticed the error. Notice how the Taylor expansion doesn't work if you use the wrong formula.

Regarding the missing `axis` argument, I think it is not needed, because `x` only has one index (the time variable `l`) and no dependence on the type of risk.

I attach the relevant part of Pintilie's book
[pintilie_var.pdf](https://github.com/user-attachments/files/18217755/pintilie_var.pdf)

sebp commented on sksurv/datasets/base.py:
```diff
@@ -474,9 +475,83 @@ def load_bmt():
+
+    The dataset has 100 samples and 4 features:::
+
+        1. dx: Diagnosis: AML=acute myeloid leukaemia, CML=chronic myeloid leukaemia, MDS=myelodysplastic syndrome
+        2. tx: Randomized treatment: BM=cell harvested from the bone marrow, PB=cell harvested from the peripheral blood
+        3. extent: Extent of disease: L=limited, E=extensive
+        5. age: Age (years)
+
+    The rest of the columns correspond to outcome variables:::
+
+        6.  survtime: Time from date of transplant to death or last follow-up (Years)
+        7.  reltime: Time from date of transplant to relapse or last follow-up (Years)
+        8.  agvhtime: Time from date of transplant to acute GVHD or last follow-up (Years)
+        9.  cgvhtime: Time from date of transplant to chronic GVHD or last follow-up (Years)
+        10. stat Status: 1=Dead, 0=Alive
+        11. rcens Relapse: 1=Yes, 0=No
+        4.  agvhdgd: Grade of acute GVHD
+        12. agvh: Acute GVHD: 1=Yes, 0=No
+        13. cgvh: Chronic GVHD: 1=Yes, 0=No
```
A [table](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#tables) might more suitable to describe the dataset. Headers could be "Index", "Name", "Description", "Encoding".

mvlvrd replied:
Done

sebp commented on sksurv/datasets/base.py:
```diff
@@ -474,9 +475,83 @@ def load_bmt():
+        3. extent: Extent of disease: L=limited, E=extensive
+        5. age: Age (years)
+
+    The rest of the columns correspond to outcome variables:::
```
The line should end with "::" for a literal block.

mvlvrd replied:
Done

sebp reviewed (changes requested):
I apologize for the delay.

I don't have any major concerns. The code seems to be correct, although I did not compare against reference implementation yet. 

I'm not sure what exactly R's cmprsk implements. mstate seems to implement the Aalen estimator. So far, I could not find an R implementation of the Dinse estimator. Only [STATA](https://journals.sagepub.com/doi/10.1177/1536867X0400400201) and [SAS](https://support.sas.com/documentation/cdl/en/statug/68162/HTML/default/viewer.htm#statug_lifetest_details25.htm) seem to have it.

I'm inclined to drop the exact version of the Dinse estimator, because the approximation seems to be more widespread (STATA and SAS only use the approximation).

mvlvrd commented:
Thanks for the review and sorry for the late response.
I have addressed most of the comments, please tell if you need more explanations or corrections.

I would like to keep the exact Dinse estimator, as it is the one I am using in production code.

mvlvrd commented on sksurv/datasets/base.py:
```diff
@@ -474,9 +475,83 @@ def load_bmt():
+    initiated for patients with a myeloid malignancy who were to
+    undergo an allogeneic bone marrow transplant.
+
+    The dataset has 100 samples and 4 features:::
```
Done

mvlvrd commented on sksurv/datasets/base.py:
```diff
@@ -474,9 +475,83 @@ def load_bmt():
+    censoring variable cens is coded as 0 when no events were
+    observed, 1 if CGVHD was observed as first event, 2 if a relapse
+    was observed as the first event and 3 if death occurred before
+    either of the events: The endpoint (status) are therefore:
```
Done

mvlvrd commented on sksurv/datasets/base.py:
```diff
@@ -474,9 +475,83 @@ def load_bmt():
+    ----------
+    .. [1] Melania Pintilie: "Competing Risks: A Practical Perspective". John Wiley & Sons, 2006
+
+    .. [2] https://drive.google.com/file/d/1FPM264pE\_-F8DB7lvFeLB1yQ3HDo7c-i/view
```
Done

mvlvrd commented on sksurv/nonparametric.py:
```diff
@@ -601,7 +601,42 @@ def predict_ipcw(self, y):
+    theta = np.zeros_like(cum_inc)
+    den = cum_inc * log_cum_i
+    np.divide(sigma_t, den, where=den < -eps, out=theta)
+    theta = z * np.multiply.outer(np.array([-1, 1]), theta)
```
`np.outer` flattens the input, (it is equivalent to `np.multiply.outer(a.ravel(), b.ravel())`) but I need to keep the dimensions of the input.
https://numpy.org/doc/2.1/reference/generated/numpy.outer.html

mvlvrd commented on sksurv/nonparametric.py:
```diff
@@ -601,7 +601,42 @@ def predict_ipcw(self, y):
+    np.log(cum_inc, where=cum_inc > eps, out=log_cum_i)
+    theta = np.zeros_like(cum_inc)
+    den = cum_inc * log_cum_i
+    np.divide(sigma_t, den, where=den < -eps, out=theta)
```
I was just following the same conventions as in `_ci_logmlog`.

I have rewritten the whole function to be more explicit:
I removed the condition `ci[:, 1.0 - cum_inc <= eps] = 1.0`, and rewrote `ci[:, cum_inc <= eps] = 0.0` as `ci[:, zero_count] = 0.0`
The reason is that for zero counts the formula for the CI is not well defined. It still gets set to (0, 0) which is not great, but I don't know of a better solution for this particular case. (I could try to think about it if needed).

sebp replied:
I 'm getting confused by my own code I guess 🤦

Can we merge `_cr_ci_logmlog()` and `_ci_logmlog()` into a single function then?

mvlvrd replied:
Thanks. I didn't notice the formulas for both cases are actually equivalent. I merged them accordingly.

mvlvrd commented on sksurv/nonparametric.py:
```diff
@@ -601,7 +601,42 @@ def predict_ipcw(self, y):
+    theta = z * np.multiply.outer(np.array([-1, 1]), theta)
+    ci = np.exp(log_cum_i * np.exp(theta))
+    ci[:, cum_inc <= eps] = 0.0
+    ci[:, 1.0 - cum_inc <= eps] = 1.0
```
See above comment.
The CI is an exponential, so with the current iteration it cannot be negative. Now it doesn't suppress the bug case of it being larger than one.

sebp replied:
I apologize, since `cum_inc` is a probability, it is valid.

mvlvrd commented on sksurv/nonparametric.py:
```diff
@@ -601,7 +601,42 @@ def predict_ipcw(self, y):
+
+
+def _cum_inc_cr_ci_estimator(cum_inc, var, conf_level, conf_type):
+    if conf_type not in {"log-log"}:
```
I decide to keep it this way to be consistent with the `_km_ci_estimator`. I can change it it you prefer it so.

mvlvrd commented on sksurv/nonparametric.py:
```diff
@@ -601,7 +601,42 @@ def predict_ipcw(self, y):
+    time_min=None,
+    conf_level=0.95,
+    conf_type=None,
+    var_type=None,
```
Done

mvlvrd commented on sksurv/nonparametric.py:
```diff
@@ -680,4 +736,103 @@ def cumulative_incidence_competing_risks(event, time_exit, time_min=None):
+    irt = cum_inc[1:, :, np.newaxis] - cum_inc[1:, np.newaxis, :]
+    mask = np.tril(np.ones_like(irt[0]))
+
+    var_a = np.einsum("rjk,jk,k->rj", irt**2, mask, dr / (n_at_risk * (n_at_risk - dr)))
```
I have rewritten all but one `einsum` to regular `sum`. I couldn't find a way to make it simpler without explicit loops or weird transpositions of the indices.

sebp commented on sksurv/nonparametric.py:
```diff
@@ -680,4 +742,103 @@ def cumulative_incidence_competing_risks(event, time_exit, time_min=None):
+    cprod = np.cumprod(1 + x) / (1 + x)
+
+    n_t = dr.size
+    i_idx = np.arange(n_t)[:, None, None]
+    j_idx = np.arange(n_t)[None, :, None]
+    k_idx = np.arange(n_t)[None, None, :]
```
Using a single instance of `np.arange(n_t)` should be sufficient.

sebp commented on sksurv/nonparametric.py:
```diff
@@ -601,7 +601,42 @@ def predict_ipcw(self, y):
+def _cr_ci_logmlog(cum_inc, sigma_t, z):
+    """Compute the pointwise log-minus-log transformed confidence intervals"""
+    eps = np.finfo(cum_inc.dtype).eps
+    zero_count = cum_inc <= eps
```
`~zero_count` is used twice. Can you define `cum_inc > eps` and negate the result only ones when setting the values to 0?

mvlvrd replied:
Done.

sebp reviewed (changes requested):
Thanks for changing the use of `einsum`. My idea was to avoid the multiplication with `mask`, but I now understand that this isn't possible with `np.cumsum`. Using `np.sum` makes it a little bit easier to understand, but it might be worth it. Using `np.einsum` should be faster due to avoiding intermediate arrays.

I would revert to using `np.einsum` and leave the `np.sum` as a comment to explain what is being computed.

mvlvrd commented on sksurv/datasets/base.py:
```diff
@@ -474,9 +475,103 @@ def load_bmt():
+    +=======+============+==============================================+===========================================+
+    | 1     | dx         | Diagnosis                                    | | AML=acute myeloid leukaemia             |
+    |       |            |                                              | | CML=chronic myeloid leukaemia           |
+    |       |            |                                              | | MDS=myelodysplastic syndrome            |
```
Sorry, I missed this. It looks like the actual data set is a subsample of the full data set:"In this book only the subgroup of 100 patients treated at the Princess Margaret Hospital are included." (Pintillie(s book Page 18)

I have removed the reference and added some comments.

sebp commented on sksurv/nonparametric.py:
```diff
@@ -188,17 +188,20 @@ def _compute_counts_truncated(event, time_enter, time_exit):
+def _ci_logmlog(s, sigma_t, z):
+    """Compute the pointwise log-minus-log transformed confidence intervals.
+    s refers to the prob_survival or the cum_inc (for the competing risks case).
+    sigma_t is the square root of the estimator of the log of the variance of s.
```
Isn't `sigma_t` the standard deviation, instead of $\sqrt{log(\widehat{\mathrm{Var}})}$?

mvlvrd replied:
I'm sorry, it is a mistake from my side. It should be `the variance of the log of s` instead of `the log of the variance of s`.

sigma_t is $\sqrt{\hat{Var}(\log{(s)})}$ because of the log-log transformation. If $\sigma =  \sqrt{\hat{Var}(s)}$ we would have to use $\exp(\sigma/(s*\log(s)))$ instead of $\exp(\sigma/log(s))$ for the CI.

sebp commented:
@mvlvrd Thanks for your hard work 🙏
